File: src/isabelle_query/render.py
"""Rendering — turning entries into the text the CLI prints.

The fourth layer of the DAG (above ``parsing``, below ``commands``).  Pure
formatting: the ``[src ...]`` extent annotation (``_format_extent``), preamble
and roadmap previews, the single-entry renderer (``render_entry``) with its
statement / verbatim / comments modes, and the shared verbosity-mode dispatch
(``_emit_matches``) that every match-listing command funnels through.

Depends only on ``model`` and ``parsing`` (``_proof_extent`` for the proof-span
walk, ``LATEX_LINE_RE`` for the figure-noise filter) — no call-graph or CLI
dependency.  The locus grammar and proof-block drill-down live with the
commands that consume them, not here.
"""

from __future__ import annotations

import re

from isabelle_query.model import CmdFlags, Entry, TheorySection
from isabelle_query.parsing import LATEX_LINE_RE, _proof_extent

def _format_extent(entry: Entry) -> str:
    """Format the `[src ...]` extent annotation for an entry.

    `src` is the entry's full span ``src_start..thy_end`` — a leading `text`
    preamble through the trailing blanks before the next entry.  The `body`
    span ``thy_line..body_end_line`` is surfaced separately whenever it is
    narrower at either end: a leading doc block (``src_start < thy_line``) or
    a trailing inter-lemma block (``body_end < thy_end``).  The body end is
    the safe cut boundary for `bin/move-block.py`; `src` is the
    end-of-region the next entry-or-section starts after.
    """
    if not entry.thy_line:
        return ""
    src_start = entry.src_start
    span_size = entry.line_count
    body_end = entry.body_end_line or entry.thy_end
    if src_start < entry.thy_line or body_end < entry.thy_end:
        body_size = body_end - entry.thy_line + 1
        return (f"[src {src_start}..{entry.thy_end}, "
                f"body {entry.thy_line}..{body_end}, "
                f"{body_size}/{span_size} lines]")
    return f"[src {src_start}..{entry.thy_end}, {span_size} lines]"


def _format_name_line(sec: TheorySection, entry: Entry) -> str:
    ext = _format_extent(entry)
    span = f" {ext}" if ext else ""
    return f"{entry.name} ({entry.tag}) — {sec.theory}{span}"


def _is_latex_noise(line: str) -> bool:
    """Lines containing LaTeX figure/typesetting markup we want to skip in
    truncated previews of text blocks (e.g. UTM.thy's tikzpicture diagrams)."""
    return bool(LATEX_LINE_RE.search(line))


def _strip_text_wrapper(lines: list[str]) -> list[str]:
    """Strip leading `text \\<open>` / `text_raw \\<open>` and trailing
    `\\<close>` from a text block body, so previews don't show the wrapper.
    Returns a copy; does nothing if the markers aren't found.
    """
    if not lines:
        return lines
    out = list(lines)
    # Strip leading "text \<open>" or "text_raw \<open>"
    first = out[0]
    m = re.match(r"^(\s*)(?:text_raw|text)\s*\\<open>\s*(.*)$", first)
    if m:
        stripped_first = (m.group(1) + m.group(2)).rstrip()
        if stripped_first:
            out[0] = stripped_first
        else:
            out = out[1:]
    if not out:
        return out
    # Strip trailing "\<close>" from last line
    last = out[-1]
    if last.rstrip().endswith("\\<close>"):
        trimmed = last.rstrip()[: -len("\\<close>")].rstrip()
        if trimmed:
            out[-1] = trimmed
        else:
            out = out[:-1]
    return out


def _truncate_preview(lines: list[str], n: int,
                      skip_latex: bool = True) -> tuple[list[str], int]:
    """Return (preview_lines, omitted_count).  Picks up to N non-blank,
    non-LaTeX content lines from the start of `lines`.  `omitted_count` is
    how many *original* lines were not included in the preview.
    """
    if n <= 0:
        return [], len(lines)
    out = []
    consumed = 0
    for line in lines:
        consumed += 1
        if not line.strip():
            continue
        if skip_latex and _is_latex_noise(line):
            continue
        out.append(line)
        if len(out) >= n:
            break
    omitted = len(lines) - consumed
    return out, max(0, omitted)


def _render_preamble(sec: TheorySection, preamble: tuple[int, int],
                     mode: str, context: int) -> str:
    """Render a preamble text block.

    mode='summary': first `context` content lines + "[+N more preamble lines]"
    mode='full':    full slice, wrapper stripped
    """
    start, end = preamble
    body = _strip_text_wrapper(sec.slice(start, end))
    block_size = len(body)
    if mode == "full":
        return "\n".join(body)
    preview, _ = _truncate_preview(body, context)
    suffix = ""
    shown = len(preview)
    remaining = block_size - shown
    if remaining == 1:
        return "\n".join(body)
    if remaining > 0:
        suffix = (f"\n  [+{remaining} more preamble lines, "
                  f"use --comments-only or -V to see]")
    return "\n".join(preview) + suffix


def _render_roadmap(roadmap: list[tuple[int, str]], context: int,
                    proof_remaining: int, mode: str) -> str:
    """Render a proof roadmap (extracted \\<comment> annotations).

    mode='summary': first `context` annotations + "...(N total of M proof lines)"
    mode='full':    all annotations
    """
    if not roadmap:
        # Fallback: show the existing "+N more proof lines" count line.
        if proof_remaining > 0:
            return (f"  [+{proof_remaining} more proof line"
                    f"{'s' if proof_remaining != 1 else ''}]")
        return ""
    if mode == "full":
        shown = roadmap
    else:
        shown = roadmap[:max(1, context)]
    out = []
    for ln, content in shown:
        out.append(f"  | line {ln}: {content}")
    if mode != "full" and len(roadmap) > len(shown):
        rest = len(roadmap) - len(shown)
        if rest == 1:
            ln, content = roadmap[len(shown)]
            out.append(f"  | line {ln}: {content}")
        else:
            out.append(f"  | ...({rest} more annotations "
                       f"in {proof_remaining}-line proof, use -U N to see more)")
    return "\n".join(out)


def _statement_text(sec: TheorySection, entry: Entry) -> str:
    """The entry's statement slice as one string: the declaration lines
    [thy_line..decl_end_line] (the lemma/def statement, not the proof).

    Falls back to `entry.text` for entries without a source location (e.g.
    an AXIOM placeholder), matching how `render_entry` degrades — so a
    statement search still sees *something* for those.
    """
    if not entry.thy_line:
        return entry.text
    return "\n".join(sec.slice(entry.thy_line, entry.decl_end_line))


def render_entry(sec: TheorySection, entry: Entry, *,
                 verbatim: bool = False,
                 statement: bool = False,
                 comments: str = "on",
                 context: int = 2) -> str:
    """Render a single entry.

    statement:       just the declaration slice [thy_line..decl_end_line]
                     (the statement, no proof) — the narrowest view
    verbatim:        full source slice [thy_line..thy_end]
    comments='on':   preamble (truncated) + header + statement + proof preview
                     + roadmap (truncated)
    comments='off':  header + statement + proof preview only (current default)
    comments='only': preamble (full) + header + roadmap (full), no statement
    context:         lines of preamble preview / roadmap entries shown

    `statement` and `verbatim` are opposite ends of the slice spectrum
    (declaration-only vs declaration+proof); `show` declares them mutually
    exclusive at the CLI.  If both somehow arrive, the narrower one wins.
    """
    ext = _format_extent(entry)
    header = f"--- {entry.name} ({entry.tag}) — {sec.theory}.thy {ext} ---"

    # No source location (e.g. AXIOM placeholder) → fall back to entry.text
    if not entry.thy_line:
        return f"{header}\n{entry.text}"

    if statement:
        body_lines = sec.slice(entry.thy_line, entry.decl_end_line)
        return header + "\n" + "\n".join(body_lines)

    if verbatim:
        body_lines = sec.slice(entry.thy_line, entry.thy_end)
        return header + "\n" + "\n".join(body_lines)

    out_parts: list[str] = []

    # Preamble (above header)
    if comments != "off" and entry.preamble:
        pmode = "full" if comments == "only" else "summary"
        rendered = _render_preamble(sec, entry.preamble, pmode, context)
        if rendered:
            pstart, pend = entry.preamble
            out_parts.append(f"--- preamble [{pstart}-{pend}] ---")
            out_parts.append(rendered)
            out_parts.append("")

    out_parts.append(header)

    if comments == "only":
        # Skip statement + proof; show only roadmap (full).
        if entry.roadmap:
            out_parts.append("--- roadmap (\\<comment> annotations) ---")
            proof_end = _proof_extent(sec, entry.proof_line, entry.thy_end) \
                if entry.proof_line else entry.thy_end
            out_parts.append(_render_roadmap(entry.roadmap, context,
                                             proof_end - entry.proof_line, "full"))
        elif not entry.preamble:
            out_parts.append("(no comment context for this entry)")
        return "\n".join(out_parts)

    # Statement + proof preview
    if entry.proof_line and entry.proof_line >= entry.decl_end_line:
        statement = sec.slice(entry.thy_line, entry.decl_end_line)
        first_proof = sec.slice(entry.proof_line, entry.proof_line)
        proof_end = _proof_extent(sec, entry.proof_line, entry.thy_end)
        remaining = max(0, proof_end - entry.proof_line)
        out_parts.append("\n".join(statement + first_proof))
        if comments != "off" and entry.roadmap:
            out_parts.append(_render_roadmap(entry.roadmap, context,
                                             remaining, "summary"))
        elif remaining == 1:
            extra = sec.slice(entry.proof_line + 1, entry.proof_line + 1)
            out_parts.append("\n".join(extra))
        elif remaining > 0:
            out_parts.append(f"  [+{remaining} more proof lines]")
    else:
        # No proof captured → just the declaration as recorded by the parser.
        body_lines = sec.slice(entry.thy_line, entry.decl_end_line)
        out_parts.append("\n".join(body_lines))

    return "\n".join(out_parts)


# ---------------------------------------------------------------------------
# Verbosity-mode dispatch
# ---------------------------------------------------------------------------

def _emit_matches(sections_by_theory: dict[str, TheorySection],
                  matches: list[Entry], pattern: str, flags: "CmdFlags",
                  *, statement: bool = False) -> None:
    # `statement` is the *render* selector (declaration-only).  It is passed
    # explicitly rather than read off `flags` so it stays a `show` concern:
    # on `find`, `flags.statement` means "match the statement slice", which
    # must not bleed into how the matched entries are rendered.
    if not matches:
        print(f"No entries matching '{pattern}'.")
        return

    if flags.mode == "count":
        print(len(matches))
        return

    if flags.mode == "names":
        for e in matches:
            print(_format_name_line(sections_by_theory[e.theory], e))
        return

    if flags.mode == "all":
        for e in matches:
            print(render_entry(sections_by_theory[e.theory], e,
                               verbatim=flags.verbatim,
                               statement=statement,
                               comments=flags.comments,
                               context=flags.context))
            print()
        return

    # mode == "first"
    e0 = matches[0]
    print(render_entry(sections_by_theory[e0.theory], e0,
                       verbatim=flags.verbatim,
                       statement=statement,
                       comments=flags.comments,
                       context=flags.context))
    if len(matches) > 1:
        print()
        print(f"[+{len(matches) - 1} more match(es).  Use --all to show, "
              f"--names for a list, --count for just the count.]")

File: query_base/src/render.scala
/*  Title:      query_base/src/render.scala

Turning entries into the text the tool prints.

Above `entries` and below `commands`: pure formatting, no index loading and no
argument parsing.  Every match-listing command funnels through `emit_matches`,
so the four verbosity modes (first / all / names / count) cannot drift apart
command by command, and every located line names its owner through
`Commands.owner_field`.

The extent annotation, the preamble and annotation previews, and the
single-entry renderer are the reference implementation's `render.py` — its
output IS the interface, so the shapes here are transcriptions, not choices.
*/

package isabelle.query


import isabelle.*

import java.nio.file.{Path => JPath}
import java.util.regex.Pattern

import scala.collection.mutable


object Render {
  /* --- qualifying a theory name far enough to name ONE theory --- */

  /* Python's `PurePath.parts`: an absolute path leads with its root, and a
     `.` component is dropped.  Written out rather than taken from
     `JPath.getName`, because the root element is exactly what makes the two
     spellings agree at the exhaustion case below. */
  def path_parts(p: JPath): List[String] = {
    val root = if (p == null) Nil else Option(p.getRoot).map(_.toString).toList
    val names =
      if (p == null) Nil
      else (0 until p.getNameCount).iterator.map(p.getName(_).toString)
        .filter(s => s.nonEmpty && s != ".").toList
    root ::: names
  }

  /* The same, for a name the USER typed: `alpha/Examples.thy` →
     `[alpha, Examples]`.  The trailing suffix is dropped from the last
     component only, which is Python's `with_suffix("")`. */
  def name_parts(name: String): List[String] = {
    val abs = name.startsWith("/")
    val segs = name.split("/", -1).toList.filter(s => s.nonEmpty && s != ".")
    val stripped = segs match {
      case Nil => Nil
      case _ =>
        val last = segs.last
        val i = last.lastIndexOf('.')
        val cut = if (i > 0 && i < last.length - 1) last.substring(0, i) else last
        segs.init :+ cut
    }
    (if (abs) List("/") else Nil) ::: stripped
  }

  /* Python's `PurePath.stem` / `.suffix`: a leading dot is not a suffix, and
     neither is a trailing one. */
  private def stem_suffix(p: JPath): (String, String) = {
    val n = p.getFileName.toString
    val i = n.lastIndexOf('.')
    if (i > 0 && i < n.length - 1) (n.substring(0, i), n.substring(i)) else (n, "")
  }

  /* `{resolved path: the shortest directory-qualified name unique here}`.

     A theory prints as its bare name, which is right for a session and wrong
     for a corpus: 461 AFP theory names are used by more than one theory,
     covering 1,219 of its 9,910 — `Examples` names nineteen different files.
     A `largest` run over the AFP was a wall of unqualified `Bla` with no way
     to tell one from another [disambig-names].

     Qualified by the MINIMAL distinguishing suffix, not by the full path.  A
     shared root prefix carries no information — every row would gain the same
     `thys/` — and the label is INPUT as well as output: `theory:line` is meant
     to round-trip, so the emitter qualifies far enough for the resolver to get
     back to one theory and no further.  `Commands.resolve_theory` matches this
     same tuple, so the two cannot drift.

     The label grows from the theory's DECLARED NAME with its directory chain
     in front, not from the file's stem.  Isabelle requires the two to agree
     on disk, and three things still follow: a section parsed from a jEdit
     BUFFER stops labelling as its scratch file; a ROOT that spells a theory
     with a directory (`theories "While/Hoare"`) labels as itself rather than
     as a bare stem no verb accepts; and collisions are decided on the name a
     reader could actually type. */
  def theory_labels(sections: List[Theory_Section]): Map[JPath, String] = {
    val by_path = mutable.LinkedHashMap.empty[JPath, List[String]]
    for (sec <- sections) {
      val p = sec.real_path
      if (!by_path.contains(p)) by_path(p) = path_parts(p.getParent) :+ sec.theory
    }
    val labels = mutable.HashMap.empty[JPath, String]
    var pending = by_path
    var depth = 1
    while (pending.nonEmpty) {
      val groups = mutable.LinkedHashMap.empty[String, mutable.ListBuffer[JPath]]
      for ((p, parts) <- pending)
        groups.getOrElseUpdate(parts.takeRight(depth).mkString("/"),
          new mutable.ListBuffer[JPath]) += p
      val nxt = mutable.LinkedHashMap.empty[JPath, List[String]]
      for ((label, paths) <- groups) {
        if (paths.length == 1) labels(paths.head) = label
        else
          for (p <- paths) {
            /* Exhausted: the paths are equal, so no suffix separates them.
               Settle rather than loop — a repeated label is a poor answer, an
               infinite loop is not an answer. */
            if (depth >= pending(p).length) labels(p) = label
            else nxt(p) = pending(p)
          }
      }
      pending = nxt
      depth += 1
    }
    labels.toMap
  }

  /* `theory_labels` re-keyed by each section's path AS STORED — the emitter's
     view.  `theory_labels` keys on the resolved path so two spellings of one
     file are one theory and one label; a located hit carries the section's
     path as the section holds it, and re-resolving that per ROW would be a
     syscall per row (`grep obtain` over the AFP prints 73,296 of them).

     A theory NAME cannot be the key: 461 AFP theory names are shared, which is
     the whole reason the label exists [disambig-loci]. */
  def locus_labels(sections: List[Theory_Section]): Map[JPath, String] = {
    val labels = theory_labels(sections)
    val out = mutable.LinkedHashMap.empty[JPath, String]
    for (sec <- sections) out(sec.path) = labels(sec.real_path)
    out.toMap
  }

  /* The `grep` / `sorry` locus for a path: its label, suffix restored.  Those
     two verbs report a FILE, because a non-`.thy` positional has no theory
     name to print — so `notes.md` stays `notes.md` as readily as `Examples.thy`
     qualifies to `alpha/Examples.thy`.  Both spellings round-trip:
     `Commands.resolve_theory` strips a trailing suffix before matching. */
  def file_locus(labels: Map[JPath, String], path: JPath): String = {
    val (stem, suffix) = stem_suffix(path)
    labels.getOrElse(path, stem) + suffix
  }

  /* The `methods` locus: the label alone, with no suffix restored — a method
     use is reported at a THEORY, the way `callers` is. */
  def theory_locus(labels: Map[JPath, String], path: JPath): String =
    labels.getOrElse(path, stem_suffix(path)._1)


  /* Lines carrying LaTeX figure / typesetting markup, skipped in a truncated
     preview of a document block: a tikzpicture is not a summary of anything. */
  val LATEX_LINE_RE: Pattern = Py.compile(
    """\\(begin|end|caption|node|draw|newlength|newcommand|settowidth|settoheight|scalebox|label)\b""")

  /* Written as code points, not as literals: the compiler's source encoding
     is not ours to assume, and these three characters are load-bearing
     output (U+2014 EM DASH, U+25B8 SMALL TRIANGLE, U+2192 RIGHTWARDS ARROW). */
  val EM_DASH: String = 0x2014.toChar.toString
  val TRIANGLE: String = 0x25b8.toChar.toString
  val ARROW: String = 0x2192.toChar.toString

  /* The enclosing locale/class as a scope step — rendered by the caller as
     `THEORY ▸ context hpk`.  An explicit `(in foo)` prints without a kind
     because the source does not say whether `foo` is a locale or a class. */
  def format_target(e: Entry): String =
    if (e.in_target.nonEmpty) "target " + e.in_target
    else e.blocks.map(kn => kn._1 + " " + kn._2).mkString(" " + TRIANGLE + " ")

  /* `[src A..B, N lines]`, widened to name the BODY span whenever that is
     narrower at either end — a leading doc block or a trailing inter-lemma
     block.  The body end is the safe cut boundary; `src` is where the next
     entry-or-section starts after. */
  def format_extent(e: Entry): String =
    if (e.thy_line == 0) ""
    else {
      val src_start = e.src_start
      val span_size = e.line_count
      val body_end = if (e.body_end_line != 0) e.body_end_line else e.thy_end
      if (src_start < e.thy_line || body_end < e.thy_end) {
        val body_size = body_end - e.thy_line + 1
        s"[src $src_start..${e.thy_end}, body ${e.thy_line}..$body_end, " +
          s"$body_size/$span_size lines]"
      }
      else s"[src $src_start..${e.thy_end}, $span_size lines]"
    }

  def format_name_line(sec: Theory_Section, e: Entry): String = {
    val ext = format_extent(e)
    val span = if (ext.nonEmpty) " " + ext else ""
    s"${e.name} (${e.tag}) $EM_DASH ${sec.theory}$span"
  }

  def is_latex_noise(line: String): Boolean = Py.found(LATEX_LINE_RE, line)

  private val TEXT_WRAPPER_RE: Pattern =
    Py.compile("""^(\s*)(?:text_raw|text)\s*\\<open>\s*(.*)$""")

  /* A document block's body without its `text \<open>` … `\<close>` wrapper,
     so a preview shows prose rather than syntax. */
  def strip_text_wrapper(lines0: List[String]): List[String] = {
    if (lines0.isEmpty) lines0
    else {
      var out = lines0
      Py.matches_at_start(TEXT_WRAPPER_RE, out.head) match {
        case Some(m) =>
          val first = Py.rstrip(m.group(1) + m.group(2))
          out = if (first.nonEmpty) first :: out.tail else out.tail
        case None =>
      }
      if (out.isEmpty) out
      else {
        val last = Py.rstrip(out.last)
        if (last.endsWith("""\<close>""")) {
          val trimmed = Py.rstrip(last.substring(0, last.length - """\<close>""".length))
          out = if (trimmed.nonEmpty) out.init :+ trimmed else out.init
        }
        out
      }
    }
  }

  /* Up to `n` non-blank, non-LaTeX lines from the start, plus how many of the
     ORIGINAL lines the preview never reached. */
  def truncate_preview(lines: List[String], n: Int, skip_latex: Boolean = true
  ): (List[String], Int) =
    if (n <= 0) (Nil, lines.length)
    else {
      val out = new mutable.ListBuffer[String]
      var consumed = 0
      val it = lines.iterator
      while (it.hasNext && out.length < n) {
        val line = it.next()
        consumed += 1
        if (Py.strip(line).nonEmpty && !(skip_latex && is_latex_noise(line))) out += line
      }
      (out.toList, (lines.length - consumed) max 0)
    }

  def render_preamble(sec: Theory_Section, preamble: (Int, Int), mode: String,
    context: Int
  ): String = {
    val (start, end) = preamble
    val body = strip_text_wrapper(sec.slice(start, end).toList)
    val block_size = body.length
    if (mode == "full") body.mkString("\n")
    else {
      val (preview, _) = truncate_preview(body, context)
      val remaining = block_size - preview.length
      if (remaining == 1) body.mkString("\n")
      else if (remaining > 0)
        preview.mkString("\n") +
          s"\n  [+$remaining more preamble lines, use --comments-only or -V to see]"
      else preview.mkString("\n")
    }
  }

  val annotation_kinds: List[String] = List("decl", "statement", "proof")

  private val kind_labels: Map[String, String] =
    Map("decl" -> "declaration", "statement" -> "statement", "proof" -> "proof")

  /* `summary`: the first `context` notes flat, beside the statement.
     `full`: every note, GROUPED by which part of the entry it annotates — in
     the dedicated prose view the grouping IS the content. */
  def render_annotations(annotations: List[(Int, String, String)], context: Int,
    proof_remaining: Int, mode: String
  ): String =
    if (annotations.isEmpty) {
      if (proof_remaining > 0)
        s"  [+$proof_remaining more proof line" + (if (proof_remaining != 1) "s" else "") + "]"
      else ""
    }
    else if (mode == "full") {
      val out = new mutable.ListBuffer[String]
      for (kind <- annotation_kinds) {
        val of_kind = annotations.filter(_._3 == kind)
        if (of_kind.nonEmpty) {
          out += "  " + kind_labels(kind) + ":"
          for ((ln, content, _) <- of_kind) out += s"    | line $ln: $content"
        }
      }
      out.mkString("\n")
    }
    else {
      val out = new mutable.ListBuffer[String]
      val shown = annotations.take(1 max context)
      for ((ln, content, _) <- shown) out += s"  | line $ln: $content"
      if (annotations.length > shown.length) {
        val rest = annotations.length - shown.length
        if (rest == 1) {
          val (ln, content, _) = annotations(shown.length)
          out += s"  | line $ln: $content"
        }
        else
          out += s"  | ...($rest more annotations in $proof_remaining-line proof, " +
            "use -U N to see more)"
      }
      out.mkString("\n")
    }

  /* The declaration slice as one string — what `find --statement` matches. */
  def statement_text(sec: Theory_Section, e: Entry): String =
    if (e.thy_line == 0) e.text
    else sec.slice(e.thy_line, e.decl_end_line).mkString("\n")

  def render_entry(sec: Theory_Section, e: Entry, verbatim: Boolean = false,
    statement: Boolean = false, comments: String = "on", context: Int = 2
  ): String = {
    val ext = format_extent(e)
    val header = s"--- ${e.name} (${e.tag}) $EM_DASH ${sec.theory}.thy $ext ---"

    if (e.thy_line == 0) header + "\n" + e.text
    else if (statement) header + "\n" + sec.slice(e.thy_line, e.decl_end_line).mkString("\n")
    else if (verbatim) header + "\n" + sec.slice(e.thy_line, e.thy_end).mkString("\n")
    else {
      val out = new mutable.ListBuffer[String]

      /* Above the header because that is where the author wrote it, and butted
         straight against it: a match LIST separates entries by one blank line,
         so an unnamed block set off the same way would read as a hit of its
         own. */
      if (comments != "off" && e.preamble.isDefined) {
        val pmode = if (comments == "only") "full" else "summary"
        val rendered = render_preamble(sec, e.preamble.get, pmode, context)
        if (rendered.nonEmpty) {
          val (pstart, pend) = e.preamble.get
          out += s"--- preamble for ${e.name} [$pstart-$pend] ---"
          out += rendered
        }
      }
      out += header

      if (comments == "only") {
        if (e.annotations.nonEmpty) {
          out += """--- annotations (\<comment>) ---"""
          out += render_annotations(e.annotations, context, 0, "full")
        }
        else if (e.preamble.isEmpty) out += "(no comment context for this entry)"
        out.mkString("\n")
      }
      else {
        if (e.proof_line > 0 && e.proof_line >= e.decl_end_line) {
          val stmt = sec.slice(e.thy_line, e.decl_end_line).toList
          /* `lemma a: "P" by simp` puts the proof ON the last statement line,
             which the statement slice has already printed. */
          val first_proof =
            if (e.proof_line <= e.decl_end_line) Nil
            else sec.slice(e.proof_line, e.proof_line).toList
          /* The same noise mask the parse used, rebuilt rather than retained:
             an `Array[Boolean]` per section would be paid by every theory in a
             resident index to answer a question only a RENDERED entry asks. */
          val src = sec.source
          val proof_end = Entries.proof_extent(src, e.proof_line, e.thy_end,
            Entries.line_mask(src.length, sec.nonisar_ranges))
          val remaining = (proof_end - e.proof_line) max 0
          out += (stmt ::: first_proof).mkString("\n")
          if (comments != "off" && e.annotations.nonEmpty)
            out += render_annotations(e.annotations, context, remaining, "summary")
          else if (remaining == 1)
            out += sec.slice(e.proof_line + 1, e.proof_line + 1).mkString("\n")
          else if (remaining > 0) out += s"  [+$remaining more proof lines]"
        }
        else {
          /* No proof captured — a `definition` lands here and its whole
             declaration is printed, so only the notes BELOW the slice are
             worth previewing. */
          out += sec.slice(e.thy_line, e.decl_end_line).mkString("\n")
          val unseen = e.annotations.filter(_._1 > e.decl_end_line)
          if (comments != "off" && unseen.nonEmpty)
            out += render_annotations(unseen, context, 0, "summary")
        }
        out.mkString("\n")
      }
    }
  }


  /* --- verbosity-mode dispatch --- */

  /* `statement` is the RENDER selector (declaration only), passed explicitly
     rather than read off the flags: on `find`, `--statement` means "match the
     statement slice" and must not bleed into how a match is displayed. */
  def emit_matches(out: Out, by_theory: Map[String, Theory_Section],
    matches: List[Entry], pattern: String, flags: Flags, statement: Boolean = false
  ): Unit = {
    /* Mode dispatch BEFORE the empty guard [count-mode-zero].  A count mode
       must print a number, and the empty case is the one a script most wants
       to branch on: with the sentence first, `$(query find X -c)` was
       arithmetic when X matched and a parse error when it did not.  `names` is
       the same rule — an empty list is the right answer for a pipeline, and a
       sentence on stdout would be read as a name.  Only the human-readable
       modes say so in words. */
    if (flags.mode == "count") out.println(matches.length.toString)
    else if (flags.mode == "names")
      for (e <- matches) out.println(format_name_line(by_theory(e.theory), e))
    else if (matches.isEmpty) out.println(s"No entries matching '$pattern'.")
    else if (flags.mode == "all")
      for (e <- matches) {
        out.println(render_entry(by_theory(e.theory), e, verbatim = flags.verbatim,
          statement = statement, comments = flags.comments, context = flags.context))
        out.println("")
      }
    else {
      val e0 = matches.head
      out.println(render_entry(by_theory(e0.theory), e0, verbatim = flags.verbatim,
        statement = statement, comments = flags.comments, context = flags.context))
      if (matches.length > 1) {
        out.println("")
        out.println(s"[+${matches.length - 1} more match(es).  Use --all to show, " +
          "--names for a list, --count for just the count.]")
      }
    }
  }
}
